lib/elixir/lib/map.ex

```diff
@@ -27,8 +30,48 @@ defmodule Map do
   @doc """
   Returns a new empty map.
   """
+  @spec new :: map
   def new, do: %{}
 
+  @doc """
+  Creates a `Map` from an enumerable.
+
+  Duplicated keys are removed, the latest one prevails.
+
+  ## Examples
+
+      iex> Map.new([{:b, 1}, {:a, 2}])
+      %{a: 2, b: 1}
+
+      iex> Map.new([a: 1, a: 2, a: 3])
+      %{a: 3}
+
+  """
+  @spec new(Enum.t) :: map
+  def new(enumerable) do
+    Enum.reduce(enumerable, %{}, fn {k, v}, acc -> put(acc, k, v) end)
```

Comment:
I am trying to optimize toward performance at expense of some redundancy. Two factors at play (1) this is a part of standard library, and (2) implementing more simple in terms of more complex goes against grain.